### ISSUE TITLE
ManyToMany join table model generates empty inputs

### ISSUE THREAD

**Issue by MichalLytek** [prisma/prisma#4004]
## Bug description

When I define models with ManyToMany relation and separate join table (it's happening when introspecting the existing database), Prisma client generates empty inputs for update many operations on that join table model:

![Code_VlTwR7sKGR](https://user-images.githubusercontent.com/10618781/96853574-cf273f00-145a-11eb-84c2-af3f600037de.png)

![Code_aInL85B5js](https://user-images.githubusercontent.com/10618781/96853576-cfbfd580-145a-11eb-9c39-9ecfc840a767.png)

## How to reproduce

Generate Prisma client from schema below.

Then try to make `updateMany` operation - you're not allowed to pass any keys to `data` object:
```ts
await prisma.studentClass.updateMany({ data: {}, where: { classId: 1 } });
```

## Expected behavior

When mutation input has no fields, the whole operation should be not available as it makes no sense.

It should be removed from DMMF so other prisma ecosystem tools don't suffer from this issue.

## Prisma information

```prisma
model Student {
  id           Int            @id @default(autoincrement())
  name         String
  StudentClass StudentClass[]
}

model Class {
  id           Int            @id @default(autoincrement())
  name         String
  StudentClass StudentClass[]
}

model StudentClass {
  studentId Int     @default(1)
  classId   Int
  student   Student @relation(fields: [studentId], references: [id])
  class     Class   @relation(fields: [classId], references: [id])

  @@id([studentId, classId])
}
```

## Environment & setup

- OS: Windows 10 x64 @ WSL2 Ubuntu 20.04 
- Database: PostgreSQL 
- Node.js version: 14.13 
- Prisma version:

```
@prisma/cli          : 2.9.0
@prisma/client       : 2.9.0
Current platform     : debian-openssl-1.1.x
Query Engine         : query-engine 369b3694b7edb869fad14827a33ad3f3f49bbc20 (at node_modules/@prisma/cli/query-engine-debian-openssl-1.1.x)
Migration Engine     : migration-engine-cli 369b3694b7edb869fad14827a33ad3f3f49bbc20 (at node_modules/@prisma/cli/migration-engine-debian-openssl-1.1.x)
Introspection Engine : introspection-core 369b3694b7edb869fad14827a33ad3f3f49bbc20 (at node_modules/@prisma/cli/introspection-engine-debian-openssl-1.1.x)
Format Binary        : prisma-fmt 369b3694b7edb869fad14827a33ad3f3f49bbc20 (at node_modules/@prisma/cli/prisma-fmt-debian-openssl-1.1.x)
Studio               : 0.296.0
Preview Features     : connectOrCreate, atomicNumberOperations
```

**Comment by pantharshit00** [comment]
I can reproduce this. I agree this is confusing for link tables to empty input. We should fix this.

**Comment by MichalLytek** [comment]
@pantharshit00 Any news about fixing that issue? 🤔

**Comment by pantharshit00** [comment]
We candidated this again for 2.19 again. Looks like this was missing the label so didn't get picked up earlier

**Comment by matthewmueller** [comment]
Folks using `typegraphql-prisma` are running into this quite a lot.

Issues:

- https://github.com/MichalLytek/typegraphql-prisma/issues/19
- https://github.com/MichalLytek/typegraphql-prisma/issues/76

**Comment by Wyctus** [comment]
Yes, @matthewmueller is right, this is very problematic for `typegraphql-prisma` users. 
Is there any news about the fix?

**Comment by rossinicolas** [comment]
Hello, there are any idea when this issue will be solved?

**Comment by jasonlimantoro** [comment]
I can confirm this is happening with any kinds of `M:N` relationships.
Right now this is blocking me from migrating from `TypeORM` to `prisma`, because it, in turn, causes a roadblock for `typegraphql-prisma`'s type generation process.

**Comment by dpetrick** [comment]
Unfortunately, we can't dedicate time to filtering out empty input types in the near future. Is it possible for you to filter these types somehow on the application layer to make it play nice with graphqljs?

The issue is not as simple as it seems because filtering empty input types implies that certain arguments (like `data`) are not always present on the input types anymore, breaking certain assertions in the query core that, while not hard to fix individually, require a great deal of diligence to get right in total.

**Comment by MichalLytek** [comment]
> The issue is not as simple as it seems because filtering empty input types implies that certain arguments (like data) are not always present on the input types anymore

That's also my concern and a hard thing to do right by myself.

I believe it should be solved on Prisma level because those empty types allows users to try to do impossible things like `updateMany` for custom N-M relations:

```ts
await prisma.studentClass.updateMany({ data: {}, where: { classId: 1 } });
```
In the example above, `data` is always of an empty object type.

**Comment by MichalLytek** [comment]
Bump! It's been a half-year since the last comment 😛
It's a pretty urgent issue I believe, and the workarounds are hard to achieve. 😞

**Comment by ZobairQ** [comment]
Bump it's a problem that I have encountered.. Are there any updates on this?

**Comment by henriqueArrazao** [comment]
Solutions?

**Comment by andrewleverette** [comment]
Has there been any movement on this? So far there has been only one model that is affected by this issue, but it would be great not to have to worry about it. The workaround that I use is to patch the model with a dummy value right after the Prisma schema is generated. I use a script that runs the command `npx prisma db pull`,  then the script uses `sed` to find the model in the schema and inject the dummy value. This works pretty well.

**Comment by padenj** [comment]
I'm hitting this issue as well generating a schema against a large legacy database with a lot of many-to-many join tables.  Are there any updates on this issue?
@andrewleverette - Interesting workaround, would you be willing to share your script that you use?

**Comment by danstarns** [comment]
I can confirm that this issue is reproducible on the `2.9.0` version you were using. Given your schema, the generated types are: 

```ts
export type StudentClassUpdateManyMutationInput = {}

export type StudentClassUpdateManyArgs = {
  data: StudentClassUpdateManyMutationInput
  where?: StudentClassWhereInput
}
```

**StudentClassUpdateManyMutationInput** is empty. 

And at runtime, the `updateMany` call throws an error: 

```
> gh-4004@1.0.0 dev /Users/danielstarns/code/clones/prisma/reproductions/reproductions/gh-4004
> pnpm ts-node ./index.ts


Invalid `prisma.studentClass.updateMany()` invocation in
/Users/danielstarns/code/clones/prisma/reproductions/reproductions/gh-4004/index.ts:61:29

{
  data: {
    studentId: 3
    ~~~~~~~~~
  }
}

Unknown arg `studentId` in data.studentId for type StudentClassUpdateManyMutationInput. Did you mean `select`? Available args:
type StudentClassUpdateManyMutationInput {

}


npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! gh-4004@1.0.0 dev: `pnpm ts-node ./index.ts`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the gh-4004@1.0.0 dev script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/danielstarns/.npm/_logs/2022-09-13T14_44_55_034Z-debug.log
```

However, if you use the latest version of Prisma the generated types are: 

**StudentClassUpdateManyMutationInput** still empty, but we now generate the **StudentClassUncheckedUpdateManyInput** type. 

```ts
export type StudentClassUpdateManyMutationInput = {}

export type StudentClassUncheckedUpdateManyInput = {
  studentId?: IntFieldUpdateOperationsInput | number
  classId?: IntFieldUpdateOperationsInput | number
}

/**
 * StudentClass updateMany
 */
export type StudentClassUpdateManyArgs = {
  /**
   * The data used to update StudentClasses.
   *
   **/
  data: XOR<StudentClassUpdateManyMutationInput, StudentClassUncheckedUpdateManyInput>
  /**
   * Filter which StudentClasses to update
   *
   **/
  where?: StudentClassWhereInput
}
```

And now longer throws a runtime error. 


I have written a reproduction here for internal use, https://github.com/prisma/reproductions/tree/main/gh-4004

For users having this issue I suggest you update to a newer version of Prisma.
